### ISSUE TITLE
why the loss and avg loss is large (train custom data)

### ISSUE THREAD

**Issue by zwqaoy** [pjreddie/darknet#347]
environment
CPU + opencv
cfg: 
yolo-voc.2.0.cfg
batch=1
subdivisions=1
filters=125  -->  filters=30
classes=1

data:
  classes= 1
  train  = NFPA_dataset/train.txt
  valid  = NFPA_dataset/test.txt
  names = data/obj.names
  backup = backup

train:
./darknet detector train cfg/obj.data cfg/yolo.2.0_obj.cfg darknet19_448.conv.23

and output:
yolo
layer     filters    size              input                output
    0 conv     32  3 x 3 / 1   416 x 416 x   3   ->   416 x 416 x  32
    1 max          2 x 2 / 2   416 x 416 x  32   ->   208 x 208 x  32
    2 conv     64  3 x 3 / 1   208 x 208 x  32   ->   208 x 208 x  64
    3 max          2 x 2 / 2   208 x 208 x  64   ->   104 x 104 x  64
    4 conv    128  3 x 3 / 1   104 x 104 x  64   ->   104 x 104 x 128
    5 conv     64  1 x 1 / 1   104 x 104 x 128   ->   104 x 104 x  64
    6 conv    128  3 x 3 / 1   104 x 104 x  64   ->   104 x 104 x 128
    7 max          2 x 2 / 2   104 x 104 x 128   ->    52 x  52 x 128
    8 conv    256  3 x 3 / 1    52 x  52 x 128   ->    52 x  52 x 256
    9 conv    128  1 x 1 / 1    52 x  52 x 256   ->    52 x  52 x 128
   10 conv    256  3 x 3 / 1    52 x  52 x 128   ->    52 x  52 x 256
   11 max          2 x 2 / 2    52 x  52 x 256   ->    26 x  26 x 256
   12 conv    512  3 x 3 / 1    26 x  26 x 256   ->    26 x  26 x 512
   13 conv    256  1 x 1 / 1    26 x  26 x 512   ->    26 x  26 x 256
   14 conv    512  3 x 3 / 1    26 x  26 x 256   ->    26 x  26 x 512
   15 conv    256  1 x 1 / 1    26 x  26 x 512   ->    26 x  26 x 256
   16 conv    512  3 x 3 / 1    26 x  26 x 256   ->    26 x  26 x 512
   17 max          2 x 2 / 2    26 x  26 x 512   ->    13 x  13 x 512
   18 conv   1024  3 x 3 / 1    13 x  13 x 512   ->    13 x  13 x1024
   19 conv    512  1 x 1 / 1    13 x  13 x1024   ->    13 x  13 x 512
   20 conv   1024  3 x 3 / 1    13 x  13 x 512   ->    13 x  13 x1024
   21 conv    512  1 x 1 / 1    13 x  13 x1024   ->    13 x  13 x 512
   22 conv   1024  3 x 3 / 1    13 x  13 x 512   ->    13 x  13 x1024
   23 conv   1024  3 x 3 / 1    13 x  13 x1024   ->    13 x  13 x1024
   24 conv   1024  3 x 3 / 1    13 x  13 x1024   ->    13 x  13 x1024
   25 route  16
   26 reorg              / 2    26 x  26 x 512   ->    13 x  13 x2048
   27 route  26 24
   28 conv   1024  3 x 3 / 1    13 x  13 x3072   ->    13 x  13 x1024
   29 conv     30  1 x 1 / 1    13 x  13 x1024   ->    13 x  13 x  30
   30 detection
mask_scale: Using default '1.000000'
Loading weights from darknet19_448.conv.23...Done!
Learning Rate: 0.001, Momentum: 0.9, Decay: 0.0005
Loaded: 0.006290 seconds
Region Avg IOU: 0.217919, Class: 1.000000, Obj: 0.889539, No Obj: 0.539300, Avg Recall: 0.000000,  count: 1
1: 299.448700, 299.448700 avg, 0.001000 rate, 21.508790 seconds, 1 images
Loaded: 0.000031 seconds
Region Avg IOU: 0.116371, Class: 1.000000, Obj: 0.461542, No Obj: 0.543704, Avg Recall: 0.000000,  count: 2
2: 298.166962, 299.320526 avg, 0.001000 rate, 21.395846 seconds, 2 images
Loaded: 0.000032 seconds
Region Avg IOU: 0.035227, Class: 1.000000, Obj: 0.768316, No Obj: 0.549684, Avg Recall: 0.000000,  count: 1
3: 317.181335, 301.106598 avg, 0.001000 rate, 21.396254 seconds, 3 images
Loaded: 0.000061 seconds
Region Avg IOU: 0.274932, Class: 1.000000, Obj: 0.205700, No Obj: 0.534559, Avg Recall: 0.000000,  count: 1
4: 282.915588, 299.287506 avg, 0.001000 rate, 21.328913 seconds, 4 images
Loaded: 0.000056 seconds
Region Avg IOU: 0.024362, Class: 1.000000, Obj: 0.934726, No Obj: 0.541315, Avg Recall: 0.000000,  count: 1
5: 372.298950, 306.588654 avg, 0.001000 rate, 21.335550 seconds, 5 images
Loaded: 0.000060 seconds
Region Avg IOU: 0.092548, Class: 1.000000, Obj: 0.221547, No Obj: 0.546854, Avg Recall: 0.000000,  count: 2
6: 305.172699, 306.447052 avg, 0.001000 rate, 21.331126 seconds, 6 images
Loaded: 0.000073 seconds
Region Avg IOU: 0.081288, Class: 1.000000, Obj: 0.947549, No Obj: 0.547968, Avg Recall: 0.000000,  count: 1
7: 311.960632, 306.998413 avg, 0.001000 rate, 21.347385 seconds, 7 images
Loaded: 0.000054 seconds
Region Avg IOU: 0.000000, Class: 1.000000, Obj: 0.600977, No Obj: 0.544880, Avg Recall: 0.000000,  count: 1
8: 303.081543, 306.606720 avg, 0.001000 rate, 21.360843 seconds, 8 images
Loaded: 0.000074 seconds
Region Avg IOU: -nan, Class: -nan, Obj: -nan, No Obj: 0.546970, Avg Recall: -nan,  count: 0
9: 286.727478, 304.618805 avg, 0.001000 rate, 21.428661 seconds, 9 images
Loaded: 0.000058 seconds
Region Avg IOU: 0.023772, Class: 1.000000, Obj: 0.634361, No Obj: 0.541323, Avg Recall: 0.000000,  count: 1
10: 323.354950, 306.492432 avg, 0.001000 rate, 21.408702 seconds, 10 images
Loaded: 0.000075 seconds
Region Avg IOU: 0.334069, Class: 1.000000, Obj: 0.961149, No Obj: 0.533974, Avg Recall: 0.000000,  count: 1
11: 282.065735, 304.049774 avg, 0.001000 rate, 21.478499 seconds, 11 images
Loaded: 0.000050 seconds
Region Avg IOU: 0.000000, Class: 1.000000, Obj: 0.117663, No Obj: 0.542661, Avg Recall: 0.000000,  count: 1
12: 285.139496, 302.158752 avg, 0.001000 rate, 21.481457 seconds, 12 images
Loaded: 0.000075 seconds
Region Avg IOU: 0.117931, Class: 1.000000, Obj: 0.802757, No Obj: 0.546877, Avg Recall: 0.000000,  count: 1
13: 302.134521, 302.156342 avg, 0.001000 rate, 21.451742 seconds, 13 images
Loaded: 0.000081 seconds
Region Avg IOU: 0.103108, Class: 1.000000, Obj: 0.698895, No Obj: 0.548117, Avg Recall: 0.000000,  count: 1
14: 300.803955, 302.021118 avg, 0.001000 rate, 21.568753 seconds, 14 images
Loaded: 0.000071 seconds
Region Avg IOU: 0.366295, Class: 1.000000, Obj: 0.993513, No Obj: 0.538940, Avg Recall: 0.000000,  count: 1
15: 290.141846, 300.833191 avg, 0.001000 rate, 22.245012 seconds, 15 images
Loaded: 0.000057 seconds

questions:

why the loss and avg loss is large

**Comment by wangwh0204** [comment]
try to change 
batch=1 -> batch=64
subdivisions=1 -> subdivisions=8

**Comment by zwqaoy** [comment]
thank you reply,  this issue is not exist  when  i use  the windows darknet version (https://github.com/AlexeyAB/darknet) . the cfg , data, image, label is same. 

 ./darknet detector train cfg/obj.data cfg/yolo-voc.2.0_obj.cfg darknet19_448.conv.23
yolo-voc
layer     filters    size              input                output
    0 conv     32  3 x 3 / 1   416 x 416 x   3   ->   416 x 416 x  32
    1 max          2 x 2 / 2   416 x 416 x  32   ->   208 x 208 x  32
    2 conv     64  3 x 3 / 1   208 x 208 x  32   ->   208 x 208 x  64
    3 max          2 x 2 / 2   208 x 208 x  64   ->   104 x 104 x  64
    4 conv    128  3 x 3 / 1   104 x 104 x  64   ->   104 x 104 x 128
    5 conv     64  1 x 1 / 1   104 x 104 x 128   ->   104 x 104 x  64
    6 conv    128  3 x 3 / 1   104 x 104 x  64   ->   104 x 104 x 128
    7 max          2 x 2 / 2   104 x 104 x 128   ->    52 x  52 x 128
    8 conv    256  3 x 3 / 1    52 x  52 x 128   ->    52 x  52 x 256
    9 conv    128  1 x 1 / 1    52 x  52 x 256   ->    52 x  52 x 128
   10 conv    256  3 x 3 / 1    52 x  52 x 128   ->    52 x  52 x 256
   11 max          2 x 2 / 2    52 x  52 x 256   ->    26 x  26 x 256
   12 conv    512  3 x 3 / 1    26 x  26 x 256   ->    26 x  26 x 512
   13 conv    256  1 x 1 / 1    26 x  26 x 512   ->    26 x  26 x 256
   14 conv    512  3 x 3 / 1    26 x  26 x 256   ->    26 x  26 x 512
   15 conv    256  1 x 1 / 1    26 x  26 x 512   ->    26 x  26 x 256
   16 conv    512  3 x 3 / 1    26 x  26 x 256   ->    26 x  26 x 512
   17 max          2 x 2 / 2    26 x  26 x 512   ->    13 x  13 x 512
   18 conv   1024  3 x 3 / 1    13 x  13 x 512   ->    13 x  13 x1024
   19 conv    512  1 x 1 / 1    13 x  13 x1024   ->    13 x  13 x 512
   20 conv   1024  3 x 3 / 1    13 x  13 x 512   ->    13 x  13 x1024
   21 conv    512  1 x 1 / 1    13 x  13 x1024   ->    13 x  13 x 512
   22 conv   1024  3 x 3 / 1    13 x  13 x 512   ->    13 x  13 x1024
   23 conv   1024  3 x 3 / 1    13 x  13 x1024   ->    13 x  13 x1024
   24 conv   1024  3 x 3 / 1    13 x  13 x1024   ->    13 x  13 x1024
   25 route  16
   26 reorg              / 2    26 x  26 x 512   ->    13 x  13 x2048
   27 route  26 24
   28 conv   1024  3 x 3 / 1    13 x  13 x3072   ->    13 x  13 x1024
   29 conv     30  1 x 1 / 1    13 x  13 x1024   ->    13 x  13 x  30
   30 detection
Loading weights from darknet19_448.conv.23...
 seen 32 
Done!
Learning Rate: 0.0001, Momentum: 0.9, Decay: 0.0005
Loaded: 0.017134 seconds
Region Avg IOU: 0.636400, Class: 1.000000, Obj: 0.376049, No Obj: 0.472032, Avg Recall: 1.000000,  count: 1
1: 11.367310, 11.367310 avg, 0.000100 rate, 20.268080 seconds, 1 images
Loaded: 0.000033 seconds
Region Avg IOU: 0.041334, Class: 1.000000, Obj: 0.232300, No Obj: 0.418250, Avg Recall: 0.000000,  count: 2
2: 13.774956, 11.608074 avg, 0.000100 rate, 20.153952 seconds, 2 images
Loaded: 0.000028 seconds
Region Avg IOU: 0.203514, Class: 1.000000, Obj: 0.342241, No Obj: 0.330646, Avg Recall: 0.000000,  count: 1
3: 7.990349, 11.246302 avg, 0.000100 rate, 20.158440 seconds, 3 images
Loaded: 0.000061 seconds
Region Avg IOU: 0.127318, Class: 1.000000, Obj: 0.067984, No Obj: 0.222316, Avg Recall: 0.000000,  count: 1
4: 4.189097, 10.540581 avg, 0.000100 rate, 20.194052 seconds, 4 images
Loaded: 0.000070 seconds
Region Avg IOU: 0.029430, Class: 1.000000, Obj: 0.046882, No Obj: 0.146149, Avg Recall: 0.000000,  count: 1
5: 1.398973, 9.626420 avg, 0.000100 rate, 20.424110 seconds, 5 images
Loaded: 0.000068 seconds
Region Avg IOU: 0.315309, Class: 1.000000, Obj: 0.112691, No Obj: 0.089732, Avg Recall: 0.000000,  count: 1

**Comment by shizukachan** [comment]
you aren't going to train yolo in 15 iterations of batch size 1. train more, pay attention to whether obj and no obj go to zero (ie. both less than 0.00001) - if so, restart training from the beginning with a lower learning rate.

also, consider using a GPU.

**Comment by corentin87** [comment]
@shizukachan i am wondering if the way of computing the loss changed for YOLOv2
I've been using the darknet repo at the commit b61bcf544e8dbcbd2e978ca6a716fa96b37df767 and was getting a loss of ~0.6.
And now i've decide to update to commit e31c50127e3d364b0b14b5a36a12ffa1a2046f44 (just before YOLOv2) and i'm getting a avg loss of ~6. Using exactly the same dataset and same cfg girl.
